File: app/tree/bfs.py
# -*- coding: utf-8 -*-
# author: Xiguang Liu<g10guang@foxmail.com>
# 2018-05-05 13:34
# 题目描述：https://www.nowcoder.com/practice/445c44d982d04483b04a54f298796288?tpId=13&tqId=11213&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


import sys

if sys.version_info < (3, 0):
    from Queue import Queue
else:
    from queue import Queue


class Solution:
    # 返回二维列表[[1,2],[4,5]]
    def Print(self, pRoot):
        result = []
        if pRoot is None:
            return result
        control = True
        queue = {True: Queue(), False: Queue()}
        queue[control].put(pRoot)
        while not queue[control].empty():
            t = []
            while not queue[control].empty():
                n = queue[control].get()
                t.append(n.val)
                if n.left:
                    queue[not control].put(n.left)
                if n.right:
                    queue[not control].put(n.right)
            control = not control
            result.append(t)
        return result

File: app/tree/bfs.go
package main

/*
 * type TreeNode struct {
 *   Val int
 *   Left *TreeNode
 *   Right *TreeNode
 * }
 */

/**
 *
 * @param pRoot TreeNode类
 * @return int整型二维数组
 */
func Print2(pRoot *TreeNode) [][]int {
	// https://www.nowcoder.com/share/jump/14730551684034623931
	var result [][]int
	var walkFn func(queue []*TreeNode)
	walkFn = func(queue []*TreeNode) {
		var line []int
		var next []*TreeNode
		for _, node := range queue {
			if node == nil {
				continue
			}

			line = append(line, node.Val)
			next = append(next, node.Left, node.Right)
		}

		if len(line) > 0 {
			result = append(result, line)
			walkFn(next)
		}

	}

	walkFn([]*TreeNode{pRoot})

	return result
}
